mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/BlockMatrix.scala

```diff
@@ -402,4 +445,402 @@ class BlockMatrix @Since("1.3.0") (
+    * @since 1.6.0
+  */
+  private [mllib] def singleBlockPLU: List[BDM[Double]] = {
+    // returns PA = LU factorization from Breeze
+    val PLU = LU(this.subBlock((0, 0), (0, 0)).toBreeze)
+    val k = PLU._1.cols
+    val L = lowerTriangular(PLU._1) - diag(diag(PLU._1)) + diag(DenseVector.fill(k){1.0})
+    val U = upperTriangular(PLU._1);
+    var P = diag(DenseVector.fill(k){1.0})
+    val Pi = diag(DenseVector.fill(k){1.0})
+    // size of square matrix
+    for(i <- 0 to (k - 1)) { // i test populating permutation matrix
+      val I = i match {case 0 => k - 1 case _ => i - 1}
+      val J = PLU._2(i) -1
+      if (i != J) {  Pi(i, J) += 1.0; Pi(J, i) += 1.0; Pi(i, i) -= 1.0; Pi(J, J) -= 1.0}
+      P = Pi * P  // constructor Pi*P for PA=LU
+      if (i != J) { Pi(i, J) -= 1.0; Pi(J, i) -= 1.0; Pi(i, i) += 1.0; Pi(J, J) += 1.0}
+    }
+    return List(P, L, U)
+  }
+
+
+  /** This method reassigns 'absolute' index locations (i,j), to sequences.  This is
+    * designed to reconsitute the orignal block locations that were lost in the
+    * subBlock method.
+    *
+    * @param rowMin The new lowest row value
+    * @param colMin The new lowest column value
+    * @return an RDD of Sequences with new block indexing
+    * @since 1.6.0
+    *
+    */
+  private [mllib] def shiftIndices(rowMin: Int, colMin: Int): RDD[((Int, Int), Matrix)] = {
+    // This routine recovers the absolute indexing of the block matrices for reassembly
+    val extractedSeq = this.blocks.map{   // shifting indices
+      case(((x, y), matrix)) => ((x + rowMin, y + colMin), matrix)
+    }
+    return extractedSeq
+  }
+
+
+
+  /** Computes the LU Decomposition of a Square Matrix.  For a matrix A of size (n x n)
+    * LU decomposition computes the Lower Triangular Matrix L, the Upper Triangular
+    * Matrix U, along with a Permutation Matrix P, such that PA=LU.  The Permutation
+    * Matrix addresses cases where zero entries prevent forward substitution
+    * solution of L or U.
+    *
+    * The BlockMatrix version takes a BlockMatrix as an input and returns a Tuple
+    * of 5 BlockMatrix objects:
+    * P, L, U (in that order), such that P.multiply(A)-L.multiply(U) = 0
+    * and Li, Ui, which are the inverse of the block diagonal terms for L and U.
+    *
+    * The blockLU method will return only P,L, and U, but blockLUtoSolver will return
+    * the extra Li and Ui matrices, which will be used by the solve method
+    * so that it does not need to recompute these values.
+    *
+    * The method follows a procedure similar to the method used in ScaLAPACK, but
+    * places more emphasis on preparing BlockMatrix objects as inputs to large
+    * BlockMatrix.multiply operations.
+    *
+    *
+    * @return P,L,U,Li,Ui as a Tuple of BlockMatrix
+    * @since 1.6.0
+  */
+
+  private [mllib] def blockLUtoSolver:
+          (BlockMatrix, BlockMatrix, BlockMatrix, BlockMatrix, BlockMatrix) = {
```

Comment:
I have not addressed this in this commit...I will do so in the next day or so for you.